File: src/main/java/com/github/sorusclient/client/ui/v1_8_9/MinecraftFontRenderer.java
package com.github.sorusclient.client.ui.v1_8_9;

import com.github.sorusclient.client.ui.IFontRenderer;
import com.github.sorusclient.client.util.Color;
import org.lwjgl.opengl.GL11;
import v1_8_9.com.mojang.blaze3d.platform.GlStateManager;
import v1_8_9.net.minecraft.client.MinecraftClient;
import v1_8_9.net.minecraft.client.font.TextRenderer;
import v1_8_9.net.minecraft.util.Identifier;

public class MinecraftFontRenderer implements IFontRenderer {

    private final TextRenderer textRenderer;

    public MinecraftFontRenderer(TextRenderer textRenderer) {
        this.textRenderer = textRenderer;
    }

    @Override
    public void drawString(String text, double x, double y, double scale, Color color) {
        MinecraftClient.getInstance().getTextureManager().bindTexture(new Identifier("assets/minecraft/textures/font/ascii.png"));
        GL11.glPushMatrix();
        GL11.glTranslated(x, y, 0);
        GL11.glScaled(scale, scale, 1);

        GlStateManager.disableDepthTest();
        GlStateManager.enableTexture();
        GlStateManager.enableBlend();

        textRenderer.draw(text, 0, 0, color.getRgb(), false);

        GlStateManager.enableDepthTest();

        GL11.glPopMatrix();
    }

    @Override
    public double getWidth(String text) {
        return textRenderer.getStringWidth(text) - 1;
    }

    @Override
    public double getHeight() {
        return textRenderer.fontHeight - 1;
    }

}

File: src/main/kotlin/com/github/sorusclient/client/adapter/v1_18_1/MinecraftFontRenderer.kt
package com.github.sorusclient.client.adapter.v1_18_1

import com.github.sorusclient.client.adapter.IFontRenderer
import com.github.sorusclient.client.adapter.IText
import com.github.sorusclient.client.util.Color
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL15
import org.lwjgl.opengl.GL20
import v1_18_1.com.mojang.blaze3d.systems.RenderSystem
import v1_18_1.net.minecraft.client.MinecraftClient
import v1_18_1.net.minecraft.client.font.TextRenderer
import v1_18_1.net.minecraft.client.util.math.MatrixStack

class MinecraftFontRenderer(private val textRenderer: TextRenderer) : IFontRenderer {

    override fun drawString(text: String, x: Double, y: Double, scale: Double, color: Color, shadow: Boolean) {
        val matrixStack = MatrixStack()
        matrixStack.scale(scale.toFloat(), scale.toFloat(), 1f)
        RenderSystem.disableDepthTest()
        RenderSystem.enableTexture()
        RenderSystem.enableBlend()

        if (shadow) {
            textRenderer.drawWithShadow(matrixStack, text, (x / scale).toFloat(), (y / scale).toFloat(), color.rgb)
        } else {
            textRenderer.draw(matrixStack, text, (x / scale).toFloat(), (y / scale).toFloat(), color.rgb)
        }

        RenderSystem.enableDepthTest()
    }

    override fun drawString(text: IText, x: Double, y: Double, scale: Double, color: Color, shadow: Boolean) {
        val matrixStack = MatrixStack()
        matrixStack.scale(scale.toFloat(), scale.toFloat(), 1f)
        RenderSystem.disableDepthTest()
        RenderSystem.enableTexture()
        RenderSystem.enableBlend()

        if (shadow) {
            textRenderer.drawWithShadow(matrixStack, Util.apiTextToText(text), (x / scale).toFloat(), (y / scale).toFloat(), color.rgb)
        } else {
            textRenderer.draw(matrixStack, Util.apiTextToText(text), (x / scale).toFloat(), (y / scale).toFloat(), color.rgb)
        }

        RenderSystem.enableDepthTest()
    }

    override fun getWidth(text: String): Double {
        return (textRenderer.getWidth(text) - 1).toDouble()
    }

    override fun getWidth(text: IText): Double {
        return (textRenderer.getWidth(Util.apiTextToText(text)) - 1).toDouble()
    }

    override fun getHeight(): Double {
        return (textRenderer.fontHeight - 1.75)
    }

}